File: 3. Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters.py
def lengthOfLongestSubstring(s):
    char_set = set()
    left = 0
    max_len = 0

    for right in range(len(s)):
        while s[right] in char_set:
            char_set.remove(s[left])
            left += 1
        char_set.add(s[right])
        max_len = max(max_len, right - left + 1)

    return max_len

if __name__ == "__main__":
    s = "abcabcbb"
    print("Input:", s)
    print("Output:", lengthOfLongestSubstring(s))  # 3

    s = "bbbbb"
    print("Input:", s)
    print("Output:", lengthOfLongestSubstring(s))  # 1

    s = "pwwkew"
    print("Input:", s)
    print("Output:", lengthOfLongestSubstring(s))  # 3

File: 3. Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters.js
function lengthOfLongestSubstring(s) {
  const set = new Set();
  let left = 0;
  let maxLen = 0;

  for (let right = 0; right < s.length; right++) {
    while (set.has(s[right])) {
      set.delete(s[left]);
      left++;
    }
    set.add(s[right]);
    maxLen = Math.max(maxLen, right - left + 1);
  }
  return maxLen;
}

function main() {
  console.log(lengthOfLongestSubstring("abcabcbb")); // 3
  console.log(lengthOfLongestSubstring("bbbbb"));    // 1
  console.log(lengthOfLongestSubstring("pwwkew"));   // 3
}

main();
